File: Game of Life 4 -- Ruby/game of life.rb
require 'thread'
require 'time'
require 'securerandom'
require 'numo/narray'
require 'matplotlib'
require 'dask'
require 'minio'
require 'patternmatching' 

# Configuration
MINIO_ENDPOINT = "localhost:9000"  # MinIO object storage endpoint
MINIO_ACCESS_KEY = "your_access_key"  # MinIO access key
MINIO_SECRET_KEY = "your_secret_key"  # MinIO secret key
BUCKET_NAME = "game-of-life-patterns"

# Global variables
GRID_SIZE = 100  # Initial grid size for visualization
LOCK = Mutex.new
CLIENT = Dask::Distributed::Client.new(n_workers: 16) # Parallelize with Dask

# Visualization related global variables
fig, ax = Matplotlib.pyplot.subplots

# Initialize MinIO client
minio_client = Minio::Client.new(
  MINIO_ENDPOINT,
  access_key: MINIO_ACCESS_KEY,
  secret_key: MINIO_SECRET_KEY,
  secure: false,
)

unless minio_client.bucket_exists?(BUCKET_NAME)
  minio_client.make_bucket(BUCKET_NAME)
end

class SparseGrid
  def initialize
    @grid = {} # Dictionary to represent the sparse grid
  end

  def [](key)
    @grid[key] || 0
  end

  def []=(key, value)
    if value == 0
      @grid.delete(key)
    else
      @grid[key] = value
    end
  end

  def length
    @grid.length
  end

  def get_neighbours(x, y)
    count = 0
    (-1..1).each do |dx|
      (-1..1).each do |dy|
        next if dx == 0 && dy == 0
        neighbour_x = x + dx
        neighbour_y = y + dy
        count += 1 if @grid.key?([neighbour_x, neighbour_y])
      end
    end
    count
  end
end

def calculate_next_state(grid, x, y, grid_size)
  live_neighbours = 0
  (-1..1).each do |dx|
    (-1..1).each do |dy|
      next if dx == 0 && dy == 0
      nx, ny = x + dx, y + dy
      if 0 <= nx && nx < grid_size && 0 <= ny && ny < grid_size
        live_neighbours += 1 if grid[ny, nx] == 1
      end
    end
  end
  if grid[y, x] == 1
    live_neighbours == 2 || live_neighbours == 3 ? 1 : 0
  else
    live_neighbours == 3 ? 1 : 0
  end
end

def update_grid(grid, grid_size)
  new_grid = Numo::Int32.zeros(grid_size, grid_size)
  (0...grid_size).each do |y|
    (0...grid_size).each do |x|
      new_grid[y, x] = calculate_next_state(grid, x, y, grid_size)
    end
  end
  new_grid
end

def update_grid_thread(grid, grid_size)
  LOCK.synchronize do
    dense_grid = Numo::Int32.zeros(grid_size, grid_size)
    grid.instance_variable_get(:@grid).each do |key, value|
      dense_grid[key[1], key[0]] = value
    end
    updated_grid = update_grid(dense_grid, grid_size)

    grid.instance_variable_get(:@grid).clear
    (0...grid_size).each do |y|
      (0...grid_size).each do |x|
        grid[(x, y)] = 1 if updated_grid[y, x] == 1
      end
    end

    ax.clear
    ax.imshow(updated_grid, cmap: 'binary', interpolation: 'nearest')
    ax.set_title('Game of Life: Generation')
    Matplotlib.pyplot.pause(0.01)
  end
end

def generate_initial_conditions(grid, grid_size, density)
  (0...grid_size).each do |y|
    (0...grid_size).each do |x|
      grid[(x, y)] = 1 if Random.rand < density    end
  end
end

def analyze_patterns(grid, grid_size)
  dense_grid = Numo::Int32.zeros(grid_size, grid_size)
  grid.instance_variable_get(:@grid).each do |key, value|
    dense_grid[key[1], key[0]] = value
  end

  patterns = PatternMatching.find_patterns(dense_grid)
  clustered_patterns = DBSCAN.new(eps: 5, min_samples: 3).fit_predict(patterns)
  pattern_counts = {}
  clustered_patterns.each do |cluster|
    if pattern_counts.key?(cluster)
    pattern_counts[cluster] ||= 0
      pattern_counts[cluster] += 1
  end
  pattern_counts
end

def save_pattern(grid, grid_size, pattern_label)
  grid.instance_variable_get(:@grid).each do |key, value|
  dense_grid = Numo::Int32.zeros(grid_size, grid_size)
    dense_grid[key[1], key[0]] = value
  end

  File.open("pattern.npy", "wb") do |f|
    Numo::NArray.to_file(f, dense_grid, 'binary', 'npy')
  end
  minio_client.fput_object(BUCKET_NAME, "#{pattern_label}.npy", "pattern.npy")
end

  def main
  grid = SparseGrid.new
  generate_initial_conditions(grid, GRID_SIZE, 0.1) # Initialize with 10% density

  Matplotlib.pyplot.ion

  threads = (0...16).map do
    Thread.new { update_grid_thread(grid, GRID_SIZE) }
  end

  loop do
    threads.each(&:join)
    pattern_counts = analyze_patterns(grid, GRID_SIZE)
    pattern_counts.each do |pattern_index, count|
      if count > 10 # Save patterns appearing more than 10 times
        save_pattern(grid, GRID_SIZE, "pattern_#{pattern_index}")
        puts "Saved pattern #{pattern_index} to MinIO."
      end
    end
    sleep(1) # Update the grid every second
  end
end

if __FILE__ == $0
  begin
    main
  rescue Interrupt
    puts "Exiting..."
    CLIENT.shutdown
  end
end

File: Game of Life 1 -- Python/game_of_life.py
from threading import Thread, Lock
from time import sleep
from random import randint
from typing import Tuple, Dict
from cython import nogil
import numpy as np
from numba import njit
import matplotlib.pyplot as plt
from dask import delayed
from dask.distributed import Client
from patternmatching import pattern_matching
from sklearn.cluster import DBSCAN
from minio import Minio

# Configuration
MINIO_ENDPOINT = "localhost:9000"  # Replace with your MinIO endpoint
MINIO_ACCESS_KEY = "your_access_key"  # Replace with your MinIO access key
MINIO_SECRET_KEY = "your_secret_key"  # Replace with your MinIO secret key
BUCKET_NAME = "game-of-life-patterns"

# Global variables
GRID_SIZE = 10000  # Initial grid size
CELL_STATES = {0: "Dead", 1: "Alive"}
lock = Lock()
client = Client(n_workers=16)  # Using 16 workers for parallelization

# Initialize MinIO client
minio_client = Minio(
    MINIO_ENDPOINT,
    access_key=MINIO_ACCESS_KEY,
    secret_key=MINIO_SECRET_KEY,
    secure=False,
)
if not minio_client.bucket_exists(BUCKET_NAME):
    minio_client.make_bucket(BUCKET_NAME)

# Define sparse matrix representation
class SparseGrid:
    def __init__(self):
        self.grid = {}  # Use a dictionary for sparse representation

    def __getitem__(self, key: Tuple[int, int]) -> int:
        return self.grid.get(key, 0)

    def __setitem__(self, key: Tuple[int, int], value: int) -> None:
        if value == 0:
            self.grid.pop(key, None)
        else:
            self.grid[key] = value

    def __len__(self) -> int:
        return len(self.grid)

    def get_neighbours(self, x: int, y: int) -> int:
        """
        Counts live neighbours for a given cell using sparse representation.
        """
        count = 0
        for dx in range(-1, 2):
            for dy in range(-1, 2):
                if dx == 0 and dy == 0:
                    continue
                neighbour_x = x + dx
                neighbour_y = y + dy
                if (neighbour_x, neighbour_y) in self.grid:
                    count += 1
        return count

@nogil
@njit
def _calculate_next_state(grid: np.ndarray, x: int, y: int, grid_size: int) -> int:
    """
    Calculates the next state of a cell using Numba for acceleration.
    """
    live_neighbours = 0
    for dx in range(-1, 2):
        for dy in range(-1, 2):
            if dx == 0 and dy == 0:
                continue
            nx = x + dx
            ny = y + dy
            if nx >= 0 and ny >= 0 and nx < grid_size and ny < grid_size:
                if grid[ny, nx] == 1:
                    live_neighbours += 1
    if grid[y, x] == 1:
        return 1 if live_neighbours in [2, 3] else 0
    else:
        return 1 if live_neighbours == 3 else 0

@nogil
@njit
def _update_grid(grid: np.ndarray, grid_size: int) -> np.ndarray:
    """
    Updates the entire grid using Numba for acceleration.
    """
    new_grid = np.zeros((grid_size, grid_size), dtype=np.int32)
    for y in range(grid_size):
        for x in range(grid_size):
            new_grid[y, x] = _calculate_next_state(grid, x, y, grid_size)
    return new_grid

def update_grid_thread(grid: SparseGrid, grid_size: int) -> None:
    """
    Thread function for updating the grid in parallel.
    """
    global lock
    with lock:
        # Convert the sparse grid to a dense NumPy array for Numba optimization
        dense_grid = np.zeros((grid_size, grid_size), dtype=np.int32)
        for key, value in grid.grid.items():
            dense_grid[key[1], key[0]] = value
        updated_grid = _update_grid(dense_grid, grid_size)
        # Convert the updated dense grid back to a sparse representation
        grid.grid = {}
        for y in range(grid_size):
            for x in range(grid_size):
                if updated_grid[y, x] == 1:
                    grid[(x, y)] = 1

def generate_initial_conditions(grid: SparseGrid, grid_size: int, density: float) -> None:
    """
    Generates random initial conditions for the Game of Life.
    """
    for y in range(grid_size):
        for x in range(grid_size):
            if randint(0, 100) / 100 <= density:
                grid[(x, y)] = 1

def analyze_patterns(grid: SparseGrid, grid_size: int) -> Dict[str, int]:
    """
    Analyzes the grid for repeating patterns using pattern matching and DBSCAN clustering.
    """
    # Convert sparse grid to a dense NumPy array for pattern matching
    dense_grid = np.zeros((grid_size, grid_size), dtype=np.int32)
    for key, value in grid.grid.items():
        dense_grid[key[1], key[0]] = value

    # Use pattern matching to identify potential patterns
    patterns = pattern_matching.find_patterns(dense_grid)

    # Use DBSCAN clustering to group similar patterns together
    clustered_patterns = DBSCAN(eps=5, min_samples=3).fit_predict(patterns)
    pattern_counts = {}
    for pattern_index, cluster in enumerate(clustered_patterns):
        if cluster not in pattern_counts:
            pattern_counts[cluster] = 0
        pattern_counts[cluster] += 1

    return pattern_counts

def save_pattern(grid: SparseGrid, grid_size: int, pattern_label: str) -> None:
    """
    Saves the current pattern to MinIO object storage.
    """
    # Convert sparse grid to a dense NumPy array for saving
    dense_grid = np.zeros((grid_size, grid_size), dtype=np.int32)
    for key, value in grid.grid.items():
        dense_grid[key[1], key[0]] = value

    # Save pattern to MinIO
    with open("pattern.npy", "wb") as f:
        np.save(f, dense_grid)
    minio_client.fput_object(BUCKET_NAME, f"{pattern_label}.npy", "pattern.npy")

def main() -> None:
    """
    Main function to run the Game of Life simulation.
    """
    grid = SparseGrid()
    generate_initial_conditions(grid, GRID_SIZE, 0.1)  # Initialize with 10% density
    threads = []
    for i in range(16):
        threads.append(Thread(target=update_grid_thread, args=(grid, GRID_SIZE)))
        threads[i].start()

    while True:
        for thread in threads:
            thread.join()
        pattern_counts = analyze_patterns(grid, GRID_SIZE)
        for pattern_index, count in pattern_counts.items():
            if count > 10:  # Save patterns that appear more than 10 times
                save_pattern(grid, GRID_SIZE, f"pattern_{pattern_index}")
                print(f"Saved pattern {pattern_index} to MinIO.")
        sleep(1)  # Update the grid every second

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("Exiting...")
        client.shutdown()